File: src/coding_agents/convert_mcp.py
"""MCP format converter — ported from convert-mcp.js to pure Python.

Reads canonical MCP server definitions from servers.json and writes
agent-specific config formats. No Node.js dependency.
"""
from __future__ import annotations

import json
import logging
from pathlib import Path

_log = logging.getLogger("coding-agents")

from coding_agents.agents import AGENTS
from coding_agents.installer.fs_ops import dry_run_mkdir
from coding_agents.utils import secure_write_text


def _merge_json(file_path: Path, data: dict) -> None:
    """Shallow-merge data into an existing JSON file (or create it)."""
    existing = {}
    if file_path.exists():
        try:
            existing = json.loads(file_path.read_text())
        except (json.JSONDecodeError, OSError):
            pass
    merged = {**existing, **data}
    secure_write_text(file_path, json.dumps(merged, indent=2) + "\n")


def _build_entry(srv: dict) -> dict:
    """Build the common MCP server entry dict from a canonical server definition."""
    entry: dict = {}
    if srv.get("command"):
        entry["command"] = srv["command"]
        entry["args"] = srv.get("args", [])
    if srv.get("url"):
        entry["url"] = srv["url"]
    if srv.get("env"):
        entry["env"] = srv["env"]
    return entry


def convert_mcp(servers_json: Path, agent_keys: list[str] | None = None) -> list[str]:
    """Convert canonical MCP servers to agent-specific formats.

    Returns list of files written.
    """
    if not servers_json.exists():
        return []

    canonical = json.loads(servers_json.read_text())
    servers = canonical.get("servers", {})
    if not servers:
        return []

    home = Path.home()
    written: list[str] = []

    if agent_keys is None:
        agent_keys = list(AGENTS.keys())

    # Dispatch table — each format maps to a writer function
    writers = {
        "claude": _write_claude,
        "codex": _write_codex,
        "pi": _write_pi,
        "opencode": lambda s, h: _write_json_mcp(s, h, h / ".config/opencode/opencode.json", "mcp"),
        "gemini": lambda s, h: _write_json_mcp(s, h, h / ".gemini/settings.json", "mcpServers"),
        "amp": lambda s, h: _write_json_mcp(s, h, h / ".config/amp/settings.json", "mcpServers"),
    }

    for key in agent_keys:
        if key not in AGENTS:
            continue
        fmt = AGENTS[key]["mcp_format"]
        writer = writers.get(fmt)
        if writer:
            _log.debug("convert_mcp: writing format=%s for agent=%s", fmt, key)
            written.extend(writer(servers, home))

    return list(set(written))


# ---------------------------------------------------------------------------
# Format-specific writers
# ---------------------------------------------------------------------------


def _write_json_mcp(
    servers: dict, home: Path, config_path: Path, top_key: str
) -> list[str]:
    """Generic writer for agents that use JSON with a top-level key for MCP servers."""
    mcp_servers = {name: _build_entry(srv) for name, srv in servers.items()}
    _merge_json(config_path, {top_key: mcp_servers})
    return [str(config_path)]


def _write_claude(servers: dict, home: Path) -> list[str]:
    """Claude Code uses mcpServers with a 'type' field."""
    mcp_servers = {}
    for name, srv in servers.items():
        entry = _build_entry(srv)
        entry["type"] = srv.get("transport", "stdio")
        mcp_servers[name] = entry

    path = home / ".mcp.json"
    _merge_json(path, {"mcpServers": mcp_servers})
    return [str(path)]


def _write_codex(servers: dict, home: Path) -> list[str]:
    """Codex CLI uses TOML with marker-based sections."""
    config_path = home / ".codex" / "config.toml"
    dry_run_mkdir(config_path.parent)

    marker_start = "# >>> coding-agents MCP >>>"
    marker_end = "# <<< coding-agents MCP <<<"

    lines = [marker_start]
    for name, srv in servers.items():
        lines.append(f"\n[mcp_servers.{name}]")
        if srv.get("command"):
            cmd_list = json.dumps([srv["command"]] + srv.get("args", []))
            lines.append(f"command = {cmd_list}")
        if srv.get("url"):
            lines.append(f'url = "{srv["url"]}"')
        if srv.get("env"):
            pairs = ", ".join(f'"{k}" = "{v}"' for k, v in srv["env"].items())
            lines.append(f"env = {{ {pairs} }}")
        lines.append("enabled = true")
    lines.append(marker_end)
    new_block = "\n".join(lines) + "\n"

    content = ""
    if config_path.exists():
        content = config_path.read_text()
        if marker_start in content:
            before = content[: content.index(marker_start)]
            after_marker = content.find(marker_end)
            after = content[after_marker + len(marker_end):] if after_marker != -1 else ""
            content = before + after

    if not content.endswith("\n"):
        content += "\n"
    content += "\n" + new_block
    secure_write_text(config_path, content)
    return [str(config_path)]


def _write_pi(servers: dict, home: Path) -> list[str]:
    """Pi uses mcp.json with settings and lifecycle: lazy."""
    mcp_servers = {}
    for name, srv in servers.items():
        entry = _build_entry(srv)
        entry["lifecycle"] = "lazy"
        mcp_servers[name] = entry

    path = home / ".pi" / "agent" / "mcp.json"
    data = {
        "settings": {"toolPrefix": "mcp", "idleTimeout": 10},
        "mcpServers": mcp_servers,
    }
    # Pi MCP is fully managed by us — overwrite
    secure_write_text(path, json.dumps(data, indent=2) + "\n")
    return [str(path)]

File: config/mcp/convert-mcp.js
#!/usr/bin/env node
/**
 * convert-mcp.js — Convert canonical MCP server definitions to agent-specific formats
 *
 * Reads: $CODING_AGENT_INSTALL_DIR/config/mcp/servers.json
 * Writes to each agent's native MCP config location.
 *
 * Canonical format (servers.json):
 * {
 *   "servers": {
 *     "server-name": {
 *       "command": "npx",
 *       "args": ["-y", "@modelcontextprotocol/server-github"],
 *       "env": { "GITHUB_TOKEN": "${GITHUB_TOKEN}" },
 *       "transport": "stdio",
 *       "url": null
 *     }
 *   }
 * }
 *
 * Sources:
 * - Claude Code .mcp.json: https://code.claude.com/docs/en/mcp
 * - Codex config.toml [mcp_servers]: https://developers.openai.com/codex/config-reference
 * - OpenCode opencode.json mcp: https://opencode.ai/docs/config/
 * - Pi mcp.json (via pi-mcp-adapter): https://github.com/nicobailon/pi-mcp-adapter
 * - Gemini CLI settings.json mcpServers: https://geminicli.com/docs/core/subagents/
 * - Amp settings: https://ampcode.com/manual
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

const home = os.homedir();
const installDir = process.env.CODING_AGENT_INSTALL_DIR || path.join(home, 'coding_agents');
const serversPath = path.join(installDir, 'config', 'mcp', 'servers.json');

if (!fs.existsSync(serversPath)) {
  console.log('No servers.json found at', serversPath);
  console.log('Create it to define shared MCP servers.');
  process.exit(0);
}

const canonical = JSON.parse(fs.readFileSync(serversPath, 'utf8'));
const servers = canonical.servers || {};

if (Object.keys(servers).length === 0) {
  console.log('No MCP servers defined in servers.json');
  process.exit(0);
}

// Helper: merge into existing JSON file
function mergeJson(filePath, data) {
  let existing = {};
  try { existing = JSON.parse(fs.readFileSync(filePath, 'utf8')); } catch {}
  const merged = { ...existing, ...data };
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(filePath, JSON.stringify(merged, null, 2) + '\n');
}

// --- Claude Code: ~/.mcp.json ---
const claudeMcp = { mcpServers: {} };
for (const [name, srv] of Object.entries(servers)) {
  claudeMcp.mcpServers[name] = {
    type: srv.transport || 'stdio',
    ...(srv.command && { command: srv.command, args: srv.args || [] }),
    ...(srv.url && { url: srv.url }),
    ...(srv.env && { env: srv.env }),
  };
}
mergeJson(path.join(home, '.mcp.json'), claudeMcp);

// --- Codex CLI: append to ~/.codex/config.toml ---
let codexToml = '\n# --- Auto-generated MCP servers (coding-agents sync) ---\n';
for (const [name, srv] of Object.entries(servers)) {
  codexToml += `\n[mcp_servers.${name}]\n`;
  if (srv.command) {
    codexToml += `command = ${JSON.stringify([srv.command, ...(srv.args || [])])}\n`;
  }
  if (srv.url) codexToml += `url = "${srv.url}"\n`;
  if (srv.env) {
    const envPairs = Object.entries(srv.env)
      .map(([k, v]) => `"${k}" = "${v}"`).join(', ');
    codexToml += `env = { ${envPairs} }\n`;
  }
  codexToml += `enabled = true\n`;
}
const codexConfig = path.join(home, '.codex', 'config.toml');
if (fs.existsSync(codexConfig)) {
  // Remove old auto-generated section
  let content = fs.readFileSync(codexConfig, 'utf8');
  content = content.replace(/\n# --- Auto-generated MCP servers.*$/s, '');
  fs.writeFileSync(codexConfig, content + codexToml);
}

// --- Pi: ~/.pi/agent/mcp.json (for pi-mcp-adapter) ---
const piMcp = {
  settings: { toolPrefix: 'mcp', idleTimeout: 10 },
  mcpServers: {},
};
for (const [name, srv] of Object.entries(servers)) {
  piMcp.mcpServers[name] = {
    ...(srv.command && { command: srv.command, args: srv.args || [] }),
    ...(srv.url && { url: srv.url }),
    ...(srv.env && { env: srv.env }),
    lifecycle: 'lazy',
  };
}
const piMcpPath = path.join(home, '.pi', 'agent', 'mcp.json');
if (!fs.existsSync(path.dirname(piMcpPath))) fs.mkdirSync(path.dirname(piMcpPath), { recursive: true });
fs.writeFileSync(piMcpPath, JSON.stringify(piMcp, null, 2) + '\n');

// --- Gemini CLI: ~/.gemini/settings.json mcpServers ---
const geminiMcp = { mcpServers: {} };
for (const [name, srv] of Object.entries(servers)) {
  geminiMcp.mcpServers[name] = {
    ...(srv.command && { command: srv.command, args: srv.args || [] }),
    ...(srv.url && { url: srv.url }),
    ...(srv.env && { env: srv.env }),
  };
}
mergeJson(path.join(home, '.gemini', 'settings.json'), geminiMcp);

// --- OpenCode: merge into opencode.json mcp section ---
const opencodeMcp = { mcp: {} };
for (const [name, srv] of Object.entries(servers)) {
  opencodeMcp.mcp[name] = {
    ...(srv.command && { command: srv.command, args: srv.args || [] }),
    ...(srv.url && { url: srv.url }),
    ...(srv.env && { env: srv.env }),
  };
}
mergeJson(path.join(home, '.config', 'opencode', 'opencode.json'), opencodeMcp);

console.log(`✅ MCP configs written for ${Object.keys(servers).length} server(s)`);
console.log('   → Claude Code (~/.mcp.json)');
console.log('   → Codex CLI (~/.codex/config.toml)');
console.log('   → Pi (~/.pi/agent/mcp.json)');
console.log('   → Gemini CLI (~/.gemini/settings.json)');
console.log('   → OpenCode (~/.config/opencode/opencode.json)');
